string_indexing/lcp_lr.py

```diff
@@ -0,0 +1,89 @@
+from common.prefix import get_longest_common_prefix
+
+def lcplr_from_lcp(lcp, n):
+  """Builds LCP-LR dictionary from LCP array"""
+  lcplr = {}
+  def fill_lcplr(left, right):
+    if right == left + 1:
+      common_prefix = lcp[right]
+    else:
+      mid = (left + right) // 2
+      common_prefix = min(fill_lcplr(left, mid), fill_lcplr(mid,right))
+    lcplr[(left, right)] = common_prefix
+    return common_prefix
+
+  fill_lcplr(1, n)
+  return lcplr
+
+
+def get_word_to_edge_lcp(l, r):
+  """Finds longest common prefix between w and one of the search bounds"""
+  return l if l >= r else r
+
+
+def get_edge_to_mid_lcp(l, r, lcplr, left, mid, right):
+  """Finds longest common prefix between mid and one of the search bounds"""
+  return lcplr[(left, mid)] if l >= r else lcplr[(mid, right)]
+
+
+def get_word_to_mid_lcp(SA, text, word, mid, word_to_edge_lcp, edge_to_mid_lcp):
+  """Finds longest common prefix between mid and w"""
+  if edge_to_mid_lcp >= word_to_edge_lcp:
+    word_to_mid_lcp = word_to_edge_lcp + \
+      get_longest_common_prefix(text[SA[mid] + word_to_edge_lcp:],
+                                word[1 + word_to_edge_lcp:])
+  else:
+    word_to_mid_lcp = edge_to_mid_lcp
+  return word_to_mid_lcp
+
+
+def initialize_lr(SA, text, word, n):
+  l = get_longest_common_prefix(text[SA[1]:], word[1:])
+  r = get_longest_common_prefix(text[SA[n]:], word[1:])
+  return l, r
+
+
+def find_bound_with_lcplr(SA, lcplr, text, word, n, m, lower_bound = True):
+  """Finds lower or upper bound of occurences of word in SA"""
+  l, r = initialize_lr(SA, text, word, n)
+
+  if lower_bound:
+    if l == m or word[l + 1:] <= text[SA[0] + l:]:
+      return 1
+    if r < m and word[r + 1:] > text[SA[n] + r:]:
+      return n + 1
+
+  else:
+    if r == m or word[r + 1:] > text[SA[n] + r:]:
+      return n + 1
```

Comment:
I didn't check carefully, but don't you need "if (something) return 1" also in the case when lower_bound == False?

Reply:
Yes, that's right. It turned out that the previous version worked, but just because there was another bug (SA[0] instead of SA[1]) and the two cancelled each other – right for a wrong reason :) Now everything should be OK